designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,224 @@
+
+Due to the fact that `NuGet.exe` is acquired standalone, it's considerably more difficult to ensure restore & build compatibility.
+
+Naturally this can lead to 2 potential different problems.
+
+#### Older NuGet tooling, newer build tooling
+
+Say a customer has pinned to a specific NuGet.exe version in their build. When a new version of Visual Studio gets released, the build machines for hosted againsts get automatically update and the customer can now run into a problem *without* any action of their own.
+
+In recent years, NuGet restore has added many capabilities to support .NET Core. From FrameworkReference in 5.3 to support for .NET 5 TFMs in 5.8.
+In *most* situations, using an older version of NuGet.exe *will not* cause any problems at build time. 5.8 and .NET SDK 5.0.100 in particular _do_ cause problems.
+[NuGet/Home#9756](https://github.com/NuGet/Home/issues/9756) introduces a change that was needed to support .NET 5 tfms and that change in itself required a significant change to how the build tasks interpret the `project.assets.json`.
+An example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/NuGet-Restore-build-issues---projectass/1190427).
+
+#### Newer NuGet tooling, older build tooling
+
+Say a customer has configured to get latest NuGet in their build tasks. While NuGet.exe and VisualStudio do ship on the same day, the rollout to all build machines is not atomic. The customers could end up in this mismatched situation without an explicit action.
+
+A newer NuGet version *might* generate an assets file that has more/different details than what the build tasks expect.
+A recent example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/vsbuild-msbuild-task-failing-since-2020-11-10/1249432) which is not 100% understood yet.
+
+In this design, we are trying to improve the general mismatched tooling experience, but also consider alternatives for the particular issues customers are hitting with VS 16.8/NuGet 5.8/ SDK 5.0.100.
+
+## Who are the customers
+
+PackageReference customers
+
+## Goals
+
+* Improve the experience for mismatched versions of restore/build tooling.
+
+## Non-Goals
+
+* Introducing explicit backwards/forward compatibility. That's a lateral effort.
+
+## Solution Overview
+
+In the 2 particular problems described above the experience when things go wrong is inconsistent. It's not *always* apparent why something goes wrong.
+This proposal covers improving that warning/error experience.
+
+### NuGet output versioning
+
+When PackageReference restore is run, NuGet writes out the 3 files mentioned above. 2 of those files have a `version`.
+
+* The `project.assets.json` has a version that has historically been used for `reading`, not tooling compatibility checks.

```

Comment:
I don't really understand what this means. If it wasn't used for compat, how was it used "for reading"?

Reply:
The version was incremented with schema changes. 

I fixed it. 

Note that it's been inconsistenly used, so we can change it to mean whatever we need it to mean if we want to :)